File: nodejs/hello-world-producer.js
/* A helloWorldProducer writes messages to a topic. */

console.log('helloWorldProducer...');

import { Kafka } from 'kafkajs';
// const { Kafka } = require('kafkajs')


const kafka = new Kafka({
    clientId: 'kafka',
    brokers: ['localhost:9092'],
})

const helloWorldProducer = kafka.producer()

await helloWorldProducer.connect()

const message = {
    value : 'Hello World from helloWorldProducer'
}

async function testProducer(message) {
    try {
        await helloWorldProducer.send({
            topic: 'test',
            messages: [
                { value: JSON.stringify(message) },
            ],
        })
        console.log(`Sent message: ${message.value}`)
    } catch (error) {
        console.error(error)
    }
}

//call the function every 5 seconds forever
setInterval(() => {
    testProducer(message)
} , 5000)

File: python/hello-world-producer.py
import json
import time
from kafka import KafkaProducer

print('helloWorldProducer...')

# Configure the Kafka producer
producer = KafkaProducer(
    bootstrap_servers=['localhost:9092'],
    value_serializer=lambda v: json.dumps(v).encode('utf-8')
)

# Define the Kafka topic
topic_name = 'test'

message = {
    'value': 'Hello World from Python Producer',
}


# Producer function
def test_producer(message):
    try:
        sent_message = producer.send(topic_name, value=message)
        metadata = sent_message.get(timeout=10)
        print(f"Sent message: {message['value']}")
        print(f"Topic: {metadata.topic}")
    except Exception as e:
        print(f"Error sending message: {str(e)}")


# Main loop to produce messages for testing
try:
    while True:
        test_producer(message)
        time.sleep(5)
except KeyboardInterrupt:
    print("Producer stopped by user")
finally:
    producer.close()
